Extending the scope of the maxFiles parameter in logging configuration

ncpibm opened the pull request:
fixes 30932 
fixes PH65488


The proposed change enhances the log cleanup mechanism in Liberty by ensuring that log files exceeding the configured maxFiles limit are deleted, regardless of whether they were created by the currently running server process or a previous instance. The existing behavior only removes old logs when the currently running process has created more than maxFiles logs, leaving older logs from previous runs untouched. As a result, when the server restarts, these old log files remain.

With the implemented logic, when the server starts, it scans the entire log output directory for files matching the log filename pattern, regardless of which process created them. It then sorts these files by their last modified timestamp, retaining only the most recent ones within the maxFiles limit and deleting the oldest ones. This ensures that the log directory does not grow indefinitely and strictly adheres to the configured retention policy. Additionally, the change includes debug-level tracing to log the deletion of old log files, improving visibility during troubleshooting.

Furthermore, in the cleanup logic, the this.backups variable is now updated before deleting old log files to ensure the newly rotated log file is properly tracked. Previously, this update occurred after the deletion logic, but repositioning it ensures the latest log file is accounted for before older ones are removed.

ncpibm committed:
updating loggeroffthread to increase the scope of maxfiles

ncpibm commented:
#build

nidhin-cp commented:
#build  ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=7cde8b3b-ce2f-4ae4-9ce9-bf3586956c3c) - :white_check_mark: completed successfully!)

_Note: Target locations of links might be accessible only to IBM employees._

ncpibm committed:
adding trace and adding the change to loggeronthread also

updating copywrite year

using poll

adding some comments and improving the logic of addbackup()

nidhin-cp commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=e4ec215e-e40d-4800-8d2e-c43a2c482f00) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

nidhin-cp commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=269d8b45-fdd3-4eec-87bd-76096e29884f) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

ncpibm committed:
test for maxfiles

pnicolucci commented on dev/io.openliberty.transport.http_fat/bootstrapping.properties:
```diff
@@ -0,0 +1,11 @@
+local.java=../../fattest.simplicity
+local.sharedLib=../../test_prereqs/moonstone/sharedlibraries
+localhost.JavaHome=/Library/Java/JavaVirtualMachines/jdk-21.0.2+13/Contents/Home
+shared.resources=../../build.sharedResources/lib
```
I don't think this is necessary, is it?

pnicolucci commented on dev/io.openliberty.transport.http_fat/fat/src/io/openliberty/transport/http_fat/AccessLogRolloverTest.java:
```diff
@@ -0,0 +1,117 @@
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
```
No need for the contributors

ncpibm replied:
ok removing contributors

pnicolucci commented on dev/io.openliberty.transport.http_fat/fat/src/io/openliberty/transport/http_fat/FATSuite.java:
```diff
@@ -10,8 +10,9 @@
  * Contributors:
  *     IBM Corporation - initial API and implementation
```
You can remove the contributors

pnicolucci commented on dev/io.openliberty.transport.http_fat/publish/files/accessLogging/server-rollover-max-backupfiles.xml:
```diff
@@ -0,0 +1,13 @@
+	terms of the Eclipse Public License 2.0 which accompanies this distribution, 
+	and is available at http://www.eclipse.org/legal/epl-2.0/ Contributors: 
+	IBM Corporation - initial API and implementation
+-->
```
Can you format this copyright better? Can remove the contributors as well:

```
<!--
    Copyright (c) 2025 IBM Corporation and others.
    All rights reserved. This program and the accompanying materials
    are made available under the terms of the Eclipse Public License 2.0
    which accompanies this distribution, and is available at
    http://www.eclipse.org/legal/epl-2.0/
    
    SPDX-License-Identifier: EPL-2.0
 -->
```

ncpibm replied:
ok

pnicolucci commented on dev/io.openliberty.transport.http_fat/publish/servers/Accesslogging/bootstrap.properties:
```diff
@@ -0,0 +1,14 @@
+# 
+# SPDX-License-Identifier: EPL-2.0
+#
+# Contributors:
```
Can remove the contributors

pnicolucci commented on dev/io.openliberty.transport.http_fat/publish/servers/Accesslogging/server.xml:
```diff
@@ -0,0 +1,13 @@
+	terms of the Eclipse Public License 2.0 which accompanies this distribution, 
+	and is available at http://www.eclipse.org/legal/epl-2.0/ Contributors: 
+	IBM Corporation - initial API and implementation
+-->
```
Please format the copyright better and remove the contributors.

pnicolucci commented on dev/io.openliberty.transport.http_fat/publish/servers/Accesslogging/common.xml:
```diff
@@ -0,0 +1,26 @@
+<server>
+
+    <featureManager>
+        <feature>ssl-1.0</feature>
```
Do we need `ssl-1.0`?

pnicolucci commented on dev/io.openliberty.transport.http_fat/publish/servers/Accesslogging/common.xml:
```diff
@@ -0,0 +1,26 @@
+    <featureManager>
+        <feature>ssl-1.0</feature>
+        <feature>servlet-3.1</feature>
+        <feature>componenttest-1.0</feature>
```
Do we need `componenttest-1.0`?

pnicolucci commented on dev/io.openliberty.transport.http_fat/publish/servers/Accesslogging/common.xml:
```diff
@@ -0,0 +1,26 @@
+    
+    <!--
+    <application type="war" id="snoop" name="snoop" location="snoop.zip" />
+    -->
```
If we're not using it remove it.

pnicolucci commented on dev/io.openliberty.transport.http_fat/publish/servers/Accesslogging/common.xml:
```diff
@@ -0,0 +1,26 @@
+    
+    <!-- use the shared/common keystore pwd: Liberty, expires 1/4/2099 -->
+    <!-- keystore is in build.sharedResources and is copied into wlp image before tests are run -->
+    <keyStore id="defaultKeyStore" password="Liberty" type="jks" location="${shared.resource.dir}/security/key.jks" />
```
Is this necessary?

pnicolucci commented on dev/io.openliberty.transport.http_fat/publish/servers/Accesslogging/common.xml:
```diff
@@ -0,0 +1,26 @@
+	This program and the accompanying materials are made available under the 
+	terms of the Eclipse Public License 2.0 which accompanies this distribution, 
+	and is available at http://www.eclipse.org/legal/epl-2.0/ Contributors: 
+	IBM Corporation - initial API and implementation
```
Format, and remove contributors

pnicolucci commented on dev/io.openliberty.transport.http_fat/fat/src/io/openliberty/transport/http_fat/AccessLogRolloverTest.java:
```diff
@@ -0,0 +1,117 @@
+
+        // Wait for log rollover to occur
+        LOG.info("Waiting for log rollover to complete...");
+        Thread.sleep(180000);
```
Is there a way to avoid these sleeps? These are going to add up.

ncpibm replied:
It is set to allow the logs to roll over. After the rollover, we will verify that our property works as expected. Another way to trigger the rollover would be to hit the URL multiple times until the logs reach the minimum file size of 1MB. However, based on my local testing, this approach would take a similar amount of time or even longer. . I will look into if there is any other way to rollover.

pnicolucci reviewed (commented):
I've given an initial review, I'd like to see the initial comments addressed and then I can take another look. Also please look into squashing your commits, I think many of them can be combined to keep the change history clean. Let me know if you have any questions.

I'd also like to see additional tests added. You're testing maxfiles = 2, which is the default, can we also test the default without specifying it? Also maybe setting maxfiles to 0 and a value other than the default of 2. What about setting an invalid value, although we're not changing this function we should try to expand the test cases in this area if at all possible.

If these tests exist someplace else let me know.

ncpibm committed:
adding xml files and adding the test cases for invalid and zero

isaacrivriv commented on dev/io.openliberty.transport.http_fat/fat/src/io/openliberty/transport/http_fat/AccessLogRolloverTest.java:
```diff
@@ -0,0 +1,280 @@
+    @Test
+    @Mode(FULL)
+    public void testmaxFilesDefaultValue() throws Exception {
+        setUp();
```
You shouldn't need to call setup in every test since you've already annotated it with Before therefore it should run before each test

ncpibm committed:
repeat the tests with Servlet 3.0 if it is available

isaacrivriv commented on dev/io.openliberty.transport.http_fat/fat/src/io/openliberty/transport/http_fat/RepeatWithServlet30.java:
```diff
@@ -0,0 +1,45 @@
+package io.openliberty.transport.http_fat;
```
Need copyright added here

isaacrivriv reviewed (commented):
A couple of changes if you can make these

ncpibm committed:
Merge branch 'integration' into 27819-Extending_scope_of_maxfiles

nidhin-cp commented:
#build
 ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=171d64da-fb66-45f6-bffb-fa8f965c4cff) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

ncpibm committed:
Merge branch 'integration' into 27819-Extending_scope_of_maxfiles

ncpibm committed:
Merge branch 'integration' into 27819-Extending_scope_of_maxfiles

isaacrivriv reviewed (approved):
LGTM, thanks for working on this!

nidhin-cp commented:
#build  ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=1bad4db5-f9cb-4f69-a6ac-31dd92fe2be8) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

mrsaldana reviewed (commented):
I recognize its been merged, just doing my review as I'm returning now from holiday. LGTM, thanks for the work here

nidhin-cp commented:
reverting the changes ,was not able to get fat approval , will be trying to get it in the next release .
